### ISSUE TITLE
Multiple From: fields using first instance for signing instead of the last.

### ISSUE THREAD

**Issue by XexDevop** [trusteddomainproject/OpenDKIM#157]
It is a common practice to append an additional "From:" header to change the sending domain from a mailer application (such as linux mail command using the -a option) This results in two From: headers, the additional (appended) one overriding the first. OpenDKIM's milter appears to use the FIRST header to determine what signature to use resulting in a DKIM check fail for non-verification of the signature. It should be parsing all the headers first, so that the additional header overrides the first and thus uses the correct sig.

**Comment by ghen2** [comment]
Common practice?  Messages with multiple From headers are not RFC 5322 compliant (the spec requires **exactly one** From header, see the table under section 3.6), so there should be no discussion about using the first or last From header... 

If you take decide to consider the last From header and declare a message compliant, but the MUA then displays the first From header to the end user, you may still allow for spoofs.

OpenDMARC has an option (RequiredHeaders) to reject incompliant messages like this, which I think is the only sensible way to deal with such messages.

**Comment by ghen2** [comment]
In particular, see https://www.rfc-editor.org/rfc/rfc6376.html#section-8.15 describing this situation.

It also provides a best practice solution for senders/signers: they can "oversign" the From header, ie. sign the existing From header plus a second null-valued one, to prevent downstream addition of extra From headers.  See "OversignHeaders" in opendkim.conf(5).

**Comment by thegushi** [comment]
I tend to agree with @ghen2 here.  Still, let me discuss things internally and see what the opinions are.

**Comment by thegushi** [comment]
Closing as by-design.

Multiple `From:` headers are not RFC 5322 compliant (§3.6 requires exactly one), and this specific scenario — appending a second `From:` to override the signed one — is documented as a known attack vector in [RFC 6376 §8.15](https://www.rfc-editor.org/rfc/rfc6376.html#section-8.15). Changing OpenDKIM to prefer the *last* `From:` header would make things worse: it would allow a downstream attacker to prepend an additional `From:` and have DKIM validate the malicious header instead.

The correct mitigation for senders who need to prevent downstream insertion of extra headers is the `OversignHeaders` option — signing the existing `From:` plus a null-valued instance, which cryptographically blocks any further additions.

Thanks @ghen2 for the thorough explanation.
